File: app/src/main/java/com/w9jds/marketbot/data/DataLoadingSubject.java
package com.w9jds.marketbot.data;

public interface DataLoadingSubject {
    boolean isDataLoading();

    void registerLoadingCallback(DataLoadingCallbacks callbacks);
    void unregisterLoadingCallback(DataLoadingCallbacks callbacks);

    void registerUpdatingCallback(DataUpdatingCallbacks callbacks);
    void unregisterUpdatingCallback(DataUpdatingCallbacks callbacks);

    interface DataLoadingCallbacks {
        void dataStartedLoading();
        void dataFinishedLoading();
        void dataFailedLoading(String errorMessage);
    }

    interface DataUpdatingCallbacks {
        void dataUpdatingStarted();
        void dataUpdatingFinished();
    }
}

File: app/src/main/java/com/w9jds/marketbot/data/DataLoadingSubject.kt
package com.w9jds.marketbot.data

interface DataLoadingSubject {
    fun isDataLoading(): Boolean

    fun registerLoadingCallback(callbacks: DataLoadingCallbacks)
    fun unregisterLoadingCallback(callbacks: DataLoadingCallbacks)

    interface DataLoadingCallbacks {
        fun dataStartedLoading()
        fun dataFinishedLoading()
        fun dataFailedLoading(errorMessage: String)
    }
}